swift/ql/lib/codeql/swift/elements/pattern/NamedPattern.qll

```diff
@@ -19,8 +19,9 @@ class NamedPattern extends Generated::NamedPattern {
    * This will be the case as long as the variable is subsequently used.
    */
   VarDecl getVarDecl() {
-    this.getImmediateEnclosingPattern*() = result.getParentPattern().getFullyUnresolved() and
-    result.getName() = this.getName()
+    this.getImmediateEnclosingPattern*() = result.getImmediateParentPattern() and
+    pragma[only_bind_out](pragma[only_bind_into](result).getName()) =
+      pragma[only_bind_out](pragma[only_bind_into](this).getName())
```

Comment:
Ugh, turns out this is fixed when quick-eval'd, but not when run as part of a query for some reason..? Is there some pragma that I'm missing? Compare this:

```
[2023-06-08 13:23:27] Evaluated non-recursive predicate NamedPattern#1696c0d8::NamedPattern::getVarDecl#ff@c38224im in 4076ms (size: 91309).
Evaluated relational algebra for predicate NamedPattern#1696c0d8::NamedPattern::getVarDecl#ff@c38224im with tuple counts:
           92048   ~0%    {3} r1 = SCAN NamedPattern#c3d26570::Generated::NamedPattern::getName#0#dispred#ff OUTPUT In.0, In.1, In.0
                      
           82893   ~0%    {2} r2 = SCAN #Pattern#19b8cf65::Pattern::getImmediateEnclosingPattern#0#dispredPlus#fb#flipped OUTPUT In.1, In.0
           66417   ~0%    {3} r3 = JOIN r2 WITH NamedPattern#c3d26570::Generated::NamedPattern::getName#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1, Lhs.0
                      
          158465   ~0%    {3} r4 = r1 UNION r3
           94246   ~2%    {3} r5 = JOIN r4 WITH VarDecl#914e0d1e::Generated::VarDecl::getImmediateParentPattern#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Rhs.1
        93510562   ~0%    {3} r6 = JOIN r5 WITH var_decls_10#join_rhs ON FIRST 1 OUTPUT Lhs.2, Rhs.1, Lhs.1
           91309   ~0%    {2} r7 = JOIN r6 WITH Synth#5f134a93::Synth::convertVarDeclToRaw#1#ff ON FIRST 2 OUTPUT Lhs.2, Lhs.0
                          return r7
```

to this:

```
[2023-06-08 13:32:59] Evaluated non-recursive predicate quick_eval#ff@fe906afo in 20ms (size: 91309).
Evaluated relational algebra for predicate quick_eval#ff@fe906afo with tuple counts:
         92048   ~0%    {3} r1 = SCAN NamedPattern#c3d26570::Generated::NamedPattern::getName#0#dispred#ff OUTPUT In.0, In.1, In.0
                    
         82893   ~0%    {2} r2 = SCAN #Pattern#19b8cf65::Pattern::getImmediateEnclosingPattern#0#dispredPlus#fb#flipped OUTPUT In.1, In.0
         66417   ~0%    {3} r3 = JOIN r2 WITH NamedPattern#c3d26570::Generated::NamedPattern::getName#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1, Lhs.0
                    
        158465   ~2%    {3} r4 = r1 UNION r3
         94246   ~3%    {3} r5 = JOIN r4 WITH VarDecl#914e0d1e::Generated::VarDecl::getImmediateParentPattern#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2
         91309   ~2%    {2} r6 = JOIN r5 WITH VarDecl#914e0d1e::Generated::VarDecl::getName#0#dispred#ff ON FIRST 2 OUTPUT Lhs.2, Lhs.0
                        return r6
```

Reply:
Never tune join orders using quick-eval. Quick-eval also evaluates `toString`, which can affect join ordering. There's an upcoming feature known as "quick-eval count" (which you might have seen some Foundations people mentioning in their stand-up) which outputs the number of results instead of the `toString` of each result. This quick-eval feature wouldn't mess with the join ordering.

Reply:
Yeah, that's what I suspected; the non-quick eval case is the normal one. Better tuple count than before, but still.